### PR TITLE
[Go SDK]: Adds Automated Python Expansion Service

### PR DESCRIPTION
**Pull request by riteshghorse**
Added an automated python expansion service for Go SDK. 

- Unit tested in this PR. 
- Integration tested on Dataframe transform at https://github.com/riteshghorse/beam/pull/52/files. Integration test will be added in a later PR once #23450 is merged. (Had to use fix beam version instead of core.SdkVersion)

Part of #23385 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by riteshghorse** [comment]
Failure are due to use of `core.SdkVersion` which points to `.dev`. Did testing with fixed version locally and it passes. Filtering the test for now.

File: sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go [hunk 1]
```diff
@@ -375,3 +380,68 @@ func jarExists(jarPath string) bool {
...
+		return "", fmt.Errorf("no python installation found: %v", err)
+	}
+	extraPackages := []string{}
+	if len(extraPackage) > 0 {
+		extraPackages = strings.Split(extraPackage, " ")
+	}
+
+	// create python virtual environment
+	sort.Strings(extraPackages)
+	beamPackage := fmt.Sprintf("apache_beam[gcp,aws,asure,dataframe]==%s", core.SdkVersion)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go:408]: Should that `asure` be `azure`, the microsoft cloud?

File: sdks/go/pkg/beam/core/runtime/xlangx/registry.go [hunk 2]
```diff
@@ -236,6 +236,8 @@ func (r *registry) getHandlerFunc(urn, expansionAddr string) (HandlerFunc, strin
 		// Leave expansionAddr unmodified so the autoNamespace keyword sticks.
 		// We strip it manually in the HandlerFunc.
 		return QueryAutomatedExpansionService, expansionAddr
+	} else if ns == autoPythonNamespace {
+		return QueryPythonExpansionService, expansionAddr
 	}
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/registry.go:241]: Style nit: use a switch instead if an if-else ladder.

```
switch ns {
 case autoJavaNamespace: 
  ...
```

File: sdks/go/pkg/beam/core/runtime/xlangx/registry.go [hunk 3]
```diff
@@ -295,6 +298,18 @@ func UseAutomatedJavaExpansionService(gradleTarget string, opts ...ExpansionServ
...
 
+// UseAutomatedPythonExpansionService takes a expansion service module name and creates a
+// tagged string to indicate that it should be used to start up an
+// automated expansion service for a cross-language expansion.
+//
+// Intended for use by cross language wrappers to permit spinning
+// up an expansion service for a user if no expansion service address
+// is provided.
+func UseAutomatedPythonExpansionService(service string) string {
+	expansionAddress := autoPythonNamespace + Separator + service
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/registry.go:309]: Remove the unnecessary expansionAddress variable, and return the concatenation directly.

File: sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go [hunk 4]
```diff
@@ -375,3 +380,68 @@ func jarExists(jarPath string) bool {
...
+	cacheDir := "~/.apache_beam/cache"
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go:409]: Put this up by the java constants,a nd build the javaJAR cache from this prefix, so we don't end up splitting our cache into different places accidentally.

**Review by lostluck — CHANGES_REQUESTED** [review]

File: sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go (hunk 1 above)
Comment (riteshghorse) [reply sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go:408]: surprisingly, there is a an asure package but I think you're right here. The Java one also uses asure https://github.com/apache/beam/blob/1c1ecb2a36dfb256a0e5eb7ff9ac5f84601b51e2/sdks/java/extensions/python/src/main/resources/org/apache/beam/sdk/extensions/python/bootstrap_beam_venv.py#L80. Need to inform them as well.

Comment (riteshghorse) [reply sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go]: typo fix PR for java is in flight as of yesterday https://github.com/apache/beam/pull/23574

File: sdks/go/pkg/beam/core/runtime/xlangx/registry.go [hunk 5]
```diff
@@ -295,6 +299,18 @@ func UseAutomatedJavaExpansionService(gradleTarget string, opts ...ExpansionServ
...
+// UseAutomatedPythonExpansionService takes a expansion service module name and creates a
+// tagged string to indicate that it should be used to start up an
+// automated expansion service for a cross-language expansion.
+//
+// Intended for use by cross language wrappers to permit spinning
+// up an expansion service for a user if no expansion service address
+// is provided.
+func UseAutomatedPythonExpansionService(service string) string {
+	return autoPythonNamespace + Separator + service
+
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/registry.go:311]: rm blank line

**Review by lostluck — APPROVED** [review]
LGTM, one tiny change.

File: sdks/go/pkg/beam/core/runtime/xlangx/registry.go (hunk 5 above)
Comment (riteshghorse) [reply sdks/go/pkg/beam/core/runtime/xlangx/registry.go]: Done

**Pull request merged**